### PR TITLE
nixos/kanidm: add basic provisioning

### PR DESCRIPTION
**Pull request by oddlama**
## Description of changes

This PR adds provisioning of persons, groups and oauth2 systems to kanidm, and allows declarative provisioning of oauth2 basic secrets and admin/idm_admin account credentials.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by h7x4** [comment]
This is really cool. Looking forward to it!

File: nixos/modules/services/security/kanidm.nix
Comment (bjornfor) [inline nixos/modules/services/security/kanidm.nix:313]: If I'm reading this right, removing a person requires marking them as `present = false`. Can you instead make it so persons that are not defined at all get removed? (I consider that the fully declarative configuration style -- which is what sets NixOS apart from e.g. Ansible.)

Comment (oddlama) [reply nixos/modules/services/security/kanidm.nix:313]: You are right you will need to specify `present = false` right now. In theory I could make it fully declarative but in practice this isn't feasible right now without writing an external tool that directly interacts with the kanidm API (which I might consider doing once the API is stable). I think it's too much work to bolt this on to the fragile and slow CLI, and this way it is at least flexible.

While I personally would like a fully declarative style here, I'm not sure that it will work properly for everyone in practice. As with all database systems you usually want to retain some kind of state and mutability at runtime. I don't know where this boundary will be and I don't know enough use cases to gauge this myself. It's hard to tell whether this would have a negative impact in some cases that I'm not aware of, especially since provisioning for other databases in nixpkgs works similarly.

Comment (bjornfor) [reply nixos/modules/services/security/kanidm.nix:313]: > I think it's too much work to bolt this on to the fragile and slow CLI, and this way it is at least flexible.

Ok. But why is fully declarative that much harder? List all users in database, remove the ones not listed in the config?

>While I personally would like a fully declarative style here, I'm not sure that it will work properly for everyone in practice.

We could have an option like `mutableUsers`? Fully declarative by default, opt-out to Ansible style declarativeness if desired, or no provisioning at all (upstream preferred way)?

Btw, I realize that https://github.com/glauth/glauth might be more aligned with "fully declarative" configuration. But kanidm seems so nice :-)

Comment (oddlama) [reply nixos/modules/services/security/kanidm.nix:313]: > Ok. But why is fully declarative that much harder?

I didn't say hard. But tedious and prone to breakage. Something I'd rather do properly.

> We could have an option like `mutableUsers`? Fully declarative by default, opt-out to Ansible style declarativeness if desired, or no provisioning at all (upstream preferred way)?

Yeah something like that. I'll think about this more once the upstream API is stable enough.

> Btw, I realize that https://github.com/glauth/glauth might be more aligned with "fully declarative" configuration. But kanidm seems so nice :-)

I don't know it, but it seems very limited compared to what kanidm can do.

**Comment by oddlama** [comment]
I've pushed some updates to this since kanidm is steadily approaching a stable release by now, and so I though this might be a good time to revisit this.

Changes:
- Removed jank. No more database editing, no more scripting using the kanidm CLI. I've written a new [provisioning utility](https://github.com/oddlama/kanidm-provision) that uses the kanidm API instead of relying on the CLI util (which should not be used for scripting according to kanidm). This is a lot faster (30s to below 1s) and more controllable in the future.
- Since provisioning secrets is explicitly against kanidm's goal of being the only source of truth, this will not be supported upstream. The only remaining option thus was to patch kanidm. I've included two optional patches here, one allows setting the admin credential when recovering accounts and the other adds an endpoint to set oauth2 basic secrets. These patches are only required when any secret should be provisioned, the regular provisioning can be used without them. When a user tries to set a secret provisioning option, an assert automatically hints at the correct option which is `services.kanidm.package = pkgs.kanidm.withSecretProvisioning;`. This was chosen so the original kanidm package is not by default affected by any patches.
- Setting `present = false;` is no longer required by default (there now an option for this called `autoRemove`). The new utility can reliably track which entities it creates and thus remove them again if needed.
- Some new options can now be provisioned, like claim maps and PKCE disable.

I've chosen to go with person, group and oauth2 provisioning for now, which suffices for classical homelab SSO and OIDC use cases. In theory we can also have unix user/group stuff, ssh key and radius provisioning in the future, but cut the scope here for now.

The NixOS tests for provisioning are not as exhaustive as I'd like them to be, so I'll address this tomorrow. Feel free to leave your suggestions here :)

File: nixos/modules/services/security/kanidm.nix
Comment (oddlama) [reply nixos/modules/services/security/kanidm.nix]: Pushed an update right now, the new utility can track entities and thus `present = false;` is no longer required. Both explicit and implicit mode are supported and can be selected by specifying `autoRemove = true/false`.

**Comment by oddlama** [comment]
Alright, I've added extensive tests now, this should be good to go. One thing I noticed is that orphaned claim maps cannot be reliably removed right now, since the required output has only been recently added to kanidm (2 weeks ago). So we need to wait for v1.1.0-rc.17 for that to work, but it doesn't impact any other functionality. I put a comment in the testing code so we can enable the test case once the next version is released.

**Review by Flakebi — APPROVED** [review]
Ran the test and tested locally, both works as expected.
Thanks, this is great!

**Comment by Flakebi** [comment]
This got some merge conflicts due to treewide doc changes

**Comment by oddlama** [comment]
Thanks for the info, i've rebased it on top of the latest master.

**Comment by erictapen** [comment]
My 5¢ here are that I'm highly skeptical of adding this kind of database provisioning to NixOS, especially with this requiring so many patches as upstream doesn't want to cover provisioning as a use case. IMO this would work better as a separate Flake.

**Comment by Flakebi** [comment]
Thanks for the rebase!

I have been using the user/group/oauth provisioning (without the patches) in the last month and I’m quite happy with how much easier everything got.
@erictapen, are you fine with adding the parts here that don’t require the patches (I think everything but secret provisioning)?

**Comment by erictapen** [comment]
@Flakebi Oh I'm fine with bringing this forward as is, iff people are enthusiastic about it. Just wanted to voice my skepticism.

**Comment by alexandru0-dev** [comment]
@oddlama love the work

About the patches rn can't be upstream as I've read already the issue that u made on the kanidm's repo.
So unless they change their minds or there is so much requests for this feature, those needs to be downstream.

Personally i think downstreaming those wouldn't be so much of a problem, tho improving clarity (using the description), warning about the fact that those patches are not associated with the kanidm project and any related issue about provisioning and patches to be reported to your repo for example.
(Better issue reporting and less burden for the kanidm team if issues are reported wrongfully there)

@erictapen @Flakebi  tell me what you think about it.

**Comment by oddlama** [comment]
Updated to support kanidm 1.2.0. Beware that the tests can currently not run correctly because the CLI logout command is broken in kanidm 1.2.0 (refer to their bug tracker for more information). I've locally verified that the tests still work when executed manually. This should fix itself by waiting i guess.

File: pkgs/by-name/ka/kanidm/package.nix
Comment (adamcstephens) [inline pkgs/by-name/ka/kanidm/package.nix:19]: I’d like to see unsupported in the name of this option. I know that’s a long word to add but we should be clear that this is something upstream fundamentally opposes.

**Review by adamcstephens — COMMENTED** [review]
Are the lib changes necessary? They add a lot of churn to this PR just shifting things around.

File: pkgs/by-name/ka/kanidm/package.nix
Comment (oddlama) [reply pkgs/by-name/ka/kanidm/package.nix:19]: Yes, this is definitely a good idea. Any ideas on how we can make sure that users know that any potential problems with provisioning should not be reported upstream?

**Comment by oddlama** [comment]
> Are the lib changes necessary? They add a lot of churn to this PR just shifting things around.

That's just how I usually structure modules when they use a lot of lib, of course it's not necessary :)
I can revert back to `lib.*` everywhere if that makes it easier to digest.

Another thing: I'm not quite sure yet on how to handle is the update process for kanidm after this change. I would like this PR to not block the regular updating process for the unpatched kanidm in any way. But the patches are currently living in nixpkgs and the nixos tests will test both the unpatched and the patched variant.

If a future update to kanidm comes along, the tests cannot succeed until the patches are updated. Does anyone have an idea on how we can decouple the tests into kanidm tests and "patched" kanidm tests that can run independently?

File: pkgs/by-name/ka/kanidm/package.nix
Comment (Flakebi) [reply pkgs/by-name/ka/kanidm/package.nix:19]: Is there any place where these options are visible except in the source here?
If it’s only here, I think a comment is fine. A link to the (closed) upstream issue would also be nice.

File: pkgs/by-name/ka/kanidm/package.nix
Comment (Flakebi) [inline pkgs/by-name/ka/kanidm/package.nix:93]: Was this for local building or is this on purpose?

Comment (oddlama) [reply pkgs/by-name/ka/kanidm/package.nix:93]: Oh. Thanks for noticing, that was by mistake and shouldn't have been included! Sorry

File: pkgs/by-name/ka/kanidm/package.nix
Comment (oddlama) [reply pkgs/by-name/ka/kanidm/package.nix:23]: I've added a comment for now, including upstream link and a notice for where to report issues with provisioning.

**Comment by oddlama** [comment]
Re-added `lib.`, added option comment, removed accidential `doCheck = false`. Rebased on master.

File: nixos/modules/services/security/kanidm.nix
Comment (Flakebi) [inline nixos/modules/services/security/kanidm.nix:493]: This option is not used

**Review by Flakebi — COMMENTED** [review]
Thanks for updating. Looks good to me, except for two unused options.

File: nixos/modules/services/security/kanidm.nix
Comment (oddlama) [reply nixos/modules/services/security/kanidm.nix:337]: Oh this was an oversight, the utility actually supports this option but I forgot to provide the flag

File: nixos/modules/services/security/kanidm.nix
Comment (oddlama) [reply nixos/modules/services/security/kanidm.nix:493]: This is not unused, like most of the options inside groups, users, .. it will be serialized to json and used by the provisioning utility

**Comment by oddlama** [comment]
> If a future update to kanidm comes along, the tests cannot succeed until the patches are updated. Does anyone have an idea on how we can decouple the tests into kanidm tests and "patched" kanidm tests that can run independently?

Any opinions on this?

**Comment by oddlama** [comment]
Rebased on master and moved release notes to 2411. Tests run successfully again now that we are on kanidm v1.2.3.

**Comment by TheRealGramdalf** [comment]
I've been using a [custom module for Kanidm](https://github.com/TheRealGramdalf/nixos/blob/566026acde5d01f8af136c02dc38a6de152964d7/mods/nixos/kani.nix) (essentially the current module with a couple tweaks), with the primary change being in the systemd service definition - all three have their type set to `notify` as per [a discussion with upstream](https://github.com/kanidm/kanidm/discussions/2892).

Should I wait to open a PR until this gets merged, or would it be better to tack it on to the end of this one somehow? If it's the former, I could also include the `lib.*` changes since everything else I'd be adding is relatively simple.

**Comment by Flakebi** [comment]
> > If a future update to kanidm comes along, the tests cannot succeed until the patches are updated. Does anyone have an idea on how we can decouple the tests into kanidm tests and "patched" kanidm tests that can run independently?
> 
> Any opinions on this?

Sorry for the slow response. Maybe the provisioning can be tested in a separate nixos test (-file)?

From a quick glance, it seems to be mostly separate and not much shared between the provisioning and the other tests.
When they are different tests, the non-patched kanidm test could still succeed when the provisioning patches don’t apply anymore.

**Comment by oddlama** [comment]
> Sorry for the slow response. Maybe the provisioning can be tested in a separate nixos test (-file)?
> 
> From a quick glance, it seems to be mostly separate and not much shared between the provisioning and the other tests. When they are different tests, the non-patched kanidm test could still succeed when the provisioning patches don’t apply anymore.

That's a good suggestion. I think the most important thing is that regular kanidm maintenance is not blocked by provisioning. Assuming I move the provisioning tests to another file, do you know a good way to expose them on the package? I'm not sure how to run them if I guard them behind the `enableProvisioning` flag without adding a package alias.

**Comment by Flakebi** [comment]
I think it’s fine to have them as another test on the package.
```nix
  passthru = {
    tests = {
      inherit (nixosTests) kanidm kanidm-provisioning;
    };
  };
```

**Comment by oddlama** [comment]
Done! I've rebased this on master to support kanidm 1.3.2, split the tests in two separate files and fixed the conflict in the release notes. This is ready to go from my side.

**Review by Flakebi — APPROVED** [review]
Thanks, looks good to me!
Tests are passing for me as well.

File: nixos/modules/services/security/kanidm.nix
Comment (adamcstephens) [inline nixos/modules/services/security/kanidm.nix:687]: I'd prefer this to be in an external systemd unit that is after and requires kanidm.service.

**Review by adamcstephens — CHANGES_REQUESTED** [review]
Can kandim-provision operate on kanidm without your patch applied? Being able to declare oauth2 resources could be useful even if one needs to manually fetch the autogenerated secrets.

The only change I want is to move the provisioning to its own systemd unit. Other than that, I'm ready to merge this. Sorry to ask you to make one more change.

**Comment by oddlama** [comment]
> Can kandim-provision operate on kanidm without your patch applied?

Yes it can. There's already a check in place that asserts a patched package is in use if any of the secret provisioning options are used.

> The only change I want is to move the provisioning to its own systemd unit. Other than that, I'm ready to merge this. Sorry to ask you to make one more change.
> I'd prefer this to be in an external systemd unit that is after and requires kanidm.service. Benefits include dedicated journal and ability to re-run it.

It took longer than expected to implement this, the decoupling proved to be more complex and problematic than anticipated. In theory, a oneshot service with `BindsTo` and `After` should be all that is needed, but this seems to have some edge cases:
- The service cannot be waited upon in test scripts unless we use `RemainsAfterExit=true` because `wait_for_unit` doesn't work with oneshot services which enter inactive state after completing.
- The service is not always started when activating a new system, even though we use `wantedBy = [ "multi-user.target" ]`. Running the test driver in interactive mode works, immediate mode doesn't. So this seems to be a hard-to-find timing issue. Sometimes it just isn't started, sometimes it is. The only workaround I found was to use `UpheldBy=kanidm.service` and `RemainsAfterExit=true` on the provisioning service, to tighten the coupling.
- Because changes to the provisioning script are not reliably picked up when switching configurations (seems to be related to oneshot services), we need two additional restart triggers on the main service. Restarting the main unit will force the bound-to provisioning to also restart.

I'm personally a bit uncomfortable with the amount of edge-cases I had to account for, because I don't fully understand why they are needed. The post-start script seemed a lot more straightforward and reliable to me, even though running the provisioning script manually is harder this way.

File: nixos/modules/services/security/kanidm.nix
Comment (oddlama) [reply nixos/modules/services/security/kanidm.nix]: See comment below

**Comment by adamcstephens** [comment]
Hmm, I wasn't expecting you to have so many issues. I've had luck in the past adding a post-service [oneshot on incus](https://github.com/NixOS/nixpkgs/blob/0da03f4801250961bc42d177711efaa4b79cec3e/nixos/modules/virtualisation/incus.nix#L395-L411). While that one is simpler since it runs as root, [the tests](https://github.com/NixOS/nixpkgs/blob/0da03f4801250961bc42d177711efaa4b79cec3e/nixos/tests/lxd/preseed.nix) haven't shown me any issues with it not working as expected.

If you want to pull it back to an ExecStartPost, go ahead and do so and we can just move forward.

Thanks!

**Comment by oddlama** [comment]
> Hmm, I wasn't expecting you to have so many issues. I've had luck in the past adding a post-service [oneshot on incus](https://github.com/NixOS/nixpkgs/blob/0da03f4801250961bc42d177711efaa4b79cec3e/nixos/modules/virtualisation/incus.nix#L395-L411). While that one is simpler since it runs as root, [the tests](https://github.com/NixOS/nixpkgs/blob/0da03f4801250961bc42d177711efaa4b79cec3e/nixos/tests/lxd/preseed.nix) haven't shown me any issues with it not working as expected.

I've had a look at incus-pressed and this is exactly what I tried too. This set of attributes (wantedBy, partOf, bindsTo, after, oneshot, RemainAfterExit) seems correct to me. Your tests run fine because you don't switch specializations. I only encounter issues when activating a new system, because either switch-to-configuration or systemd is not doing the correct thing here. My guess is that if you were to add incus-pressed to a system that is already running incus by setting `cfg.pressed` after the fact, the service would not be started when switching to the new configuration - assuming there is no configuration change to the main unit which would cause it to restart.

I know there are some issues with `PartOf` when restarts happen on the main unit while the added unit is still activating (https://github.com/systemd/systemd/issues/13841), but this should not be the case here. My best guess is that there's a bug in switch-to-configuration.pl or I misunderstand how systemd works here.
 
> If you want to pull it back to an ExecStartPost, go ahead and do so and we can just move forward.

I think that is the better choice for the time being, I've reverted the change. If we figure out what's the culprit later we can still update the unit in a separate PR. Should be a non-breaking change.

Thanks for your patience

**Review by adamcstephens — APPROVED** [review]

**Pull request merged**
